common/archiver/s3store/visibility_archiver.go

```diff
@@ -209,12 +211,54 @@ func (v *visibilityArchiver) queryAll(
-		pageSize:      request.PageSize,
-		nextPageToken: request.NextPageToken,
-		parsedQuery:   &parsedQuery{},
-	}, saTypeMap, constructVisibilitySearchPrefix(uri.Path(), request.NamespaceID))
+	// remaining is the number of workflow executions left to return before we reach pageSize.
+	remaining := request.PageSize
+	nextPageToken := request.NextPageToken
+	var executions []*workflowpb.WorkflowExecutionInfo
+	// We need to loop because the number of workflow executions returned by each call to query may be fewer than
+	// pageSize. This is because we may have to skip some workflow executions after querying S3 because there are
+	// 4 entries in S3 for each workflow execution (one for each combination of primary and secondary index), and we
+	// only want to return one entry per workflow execution. See createIndexesToArchive for a list of indexes.
+	for {
+		searchPrefix := constructVisibilitySearchPrefix(uri.Path(), request.NamespaceID)
+		// The pageSize we supply here is actually the maximum number of keys to fetch from S3. For each execution,
+		// there should be 4 keys in S3, so you might think that we should multiply the pageSize by 4. However, if we
+		// do that, we may end up returning more than pageSize workflow executions to the end user of this API. This is
+		// because we aren't guaranteed that all 4 keys for a given workflow execution will be returned in the same
+		// call. For example, if the user supplies a pageSize of 1, and we specify a maximum number of keys of 4 to S3,
+		// we may get back entries from S3 for 2 different workflow executions. You might think that we can just
+		// truncate this result to 1 workflow execution, but then the nextPageToken would be incorrect. So, we may need
+		// to make multiple calls to S3 to get the correct number of workflow executions, which will probably make this
+		// API call very slow.
+		res, err := v.queryPrefix(ctx, uri, &queryVisibilityRequest{
+			namespaceID:   request.NamespaceID,
+			pageSize:      remaining,
+			nextPageToken: nextPageToken,
+			parsedQuery:   &parsedQuery{},
+		}, saTypeMap, searchPrefix, isWorkflowTypeAndCloseTimeoutEntry)
+		if err != nil {
+			return nil, err
+		}
+		nextPageToken = res.NextPageToken
+		executions = append(executions, res.Executions...)
+		remaining -= len(res.Executions)
+		if len(nextPageToken) == 0 || remaining <= 0 {
+			break
+		}
+	}
+	return &archiver.QueryVisibilityResponse{
+		Executions:    executions,
+		NextPageToken: nextPageToken,
+	}, nil
+}
+
+// isWorkflowTypeAndCloseTimeoutEntry returns true if the S3 key is for the entry that indexes the execution by both
+// workflow type and close timeout. One example key is:
+//
+//	visibility/workflowTypeName/Workflow/closeTimeout/2023-08-01T20:51:03Z/a9c567eb-d034-4f73-a895-86fcdddc1089
+func isWorkflowTypeAndCloseTimeoutEntry(key string) bool {
+	return strings.Contains(key, primaryIndexKeyWorkflowTypeName) &&
+		strings.Contains(key, secondaryIndexKeyCloseTimeout)
```

Comment:
Shouldn't this be
```suggestion
	return strings.Contains(key, "/"+primaryIndexKeyWorkflowTypeName+"/") &&
		strings.Contains(key, "/"+secondaryIndexKeyCloseTimeout+"/")
```
to correctly account for cases where the workflow type name contains the string `closeTimeout`?

Also, is there anything preventing slashes from appearing in workflowTypeName? I guess that would be pretty weird, and probably nothing we can do about it now.

Reply:
I improved this to do some more structural parsing, splitting on "/" now.